[시온] 미션3: 체스판 초기화 제출합니다.

ehdrhelr opened the pull request:
안녕하십니까.

코드스쿼드 2021 마스터즈 백엔드 과정 Shion 입니다.

- Board 클래스에서 흰색 pawn과 검은색 pawn을 각각 다른 list에서 관리하기 시작합니다.
- 2차원 배열을 사용해서 체스판을 표현합니다.
- 메소드가 하나의 기능을 하도록 노력했습니다.
- 하드코딩을 지양하기 위해 상수를 사용했습니다.

미션3 코드 리뷰 부탁드립니다.

감사합니다!

Commit:
feat : add a String variable, representation, in the Pawn class

Add literals for it, a constructor using it and the getter.

The PawnTest class uses an added parameter.

Commit:
feat : add initializing chess board and printing it

1. Board클래스에서 black pawn과 white pawn을 각각 다른 list에서 관리하기 시작했습니다.
2. 2차원 문자열 배열로 체스판을 표현합니다.
3. "."으로 빈 체스판을 표현하고 pawn도 위치에 맞게 배열에 넣어준 후 출력합니다.
4. 이에 대한 테스트 코드를 작성했습니다.
6. 하드코딩을 지양하기 위해 상수를 정의했습니다.
5. BoardTest클래스에서 기존에 작성된 테스트 코드들은 주석처리해두었습니다. 코드 리뷰를 반영하여 수정할 예정입니다.

Commit:
feat : add the ChessGame class

1. ChessGame클래스는 main메소드를 갖고 사용자의 입력에 따라 시작과 종료 기능을 제공합니다.

Commit:
test : refactor test

미션 3의 변경사항과 미션 2의 코드 리뷰를 반영하여 테스트 코드를 작성하였습니다.

Commit:
merge upstream ehdrhelr

wheejuni commented on src/main/java/chess/ChessGame.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private void start() {
+        String[][] initializedBoard = board.initialize();
```
요구사항을 다소 잘못 이해하신 것 같습니다.
`.initialize()`의 결과물은, `String` 으로 리턴되기 이전에 `Board` 가 자기 자신의 상태를 관리하도록 구현되어야 합니다. 
`Board` 가 자신의 상태를 관리하고 있다가, 자신이 관리하던 상태를 `.print()` 로 표현하는 것입니다. 
따라서 `.initialize()` 의 결과를 외부에서 관리하면 안됩니다.
유념하시고 구현의 방향을...조금 많이 수정해주셔야 할 것 같습니다.

ehdrhelr replied:
리뷰 감사드립니다.
저도 뭔가 찝찝했는데 이유를 알지 못해서 대수롭지 않게 넘겼던 것 같습니다.
브라이언 리뷰 덕분에 뭐가 잘못됐는지 정확히 알게 된 것 같아서 속이 시원하네요ㅎㅎㅎ
말씀해주신대로 개선해보겠습니다!

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -3,26 +3,114 @@
+    }
+
+    /* 빈 체스판을 표현하기 위해 board에 점들을 채운다. */
+    private String[][] fillDotsOnBoard() {
```
앞으로 **모든 과제에서** 2차원 배열의 사용은 불가피한 경우를 제외하곤 삼가 주세요.
사실 불가피한 경우란 없습니다. 2차원 배열의 사용을 하지 않는 것을 제안드립니다. 
어떻게 하면 2차원 배열을 사용하지 않고 코딩할 수 있을까요? 체스판을 실제로 보신 일이 있나요? 이것을 추상화하려면 어떻게 표현할 수 있을까요?

ehdrhelr replied:
알고리즘 문제를 풀다가 2D맵을 2차원배열로 표현하는 것에 착안해서 사용하게 되었는데요!
제가 아는 선에서 2차원 배열로 표현하는 것도 썩 괜찮은 방법이라고 생각했는데(사실 아는게 이것밖에 없습니다..🥲), 혹시 사용하지 말라고 하시는 이유가 있을까요??
2차원 배열을 사용하지 않는다면 한 라인씩 String으로 append해서 체스판을 표현해야 할 것 같습니다...!

wheejuni replied:
좀 더 구체적인 추상화(말이 이상하군요)를 기대했는데, 생각해보니 지금 단계에서 고민할 것은 아닌 것 같습니다.
우선 3단계에서의 논의는 여기까지로 하죠.

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -3,26 +3,114 @@
+
+    /* 빈 체스판을 표현하기 위해 board에 점들을 채운다. */
+    private String[][] fillDotsOnBoard() {
+        String[][] board = new String[Board.HEIGHT][Board.WIDTH];  // 편의를 위해 index 0은 사용하지 않는다. ( 1 ~ 8 사용)
```
_편의를 위해 index 0을 사용하지 않는다_ 는 건 전혀 이해되지 않습니다. 
조금만 더 설명해주세요. 왜 사용하지 않는거죠.

ehdrhelr replied:
체스판 열이 1부터 시작하기에 직관적으로 index 1부터 사용했습니다!

주석을 자세하게 작성하지 않아 의도 전달이 잘 되지 않은 것 같네요ㅜㅜ

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -3,26 +3,114 @@
+    private String[][] fillDotsOnBoard() {
+        String[][] board = new String[Board.HEIGHT][Board.WIDTH];  // 편의를 위해 index 0은 사용하지 않는다. ( 1 ~ 8 사용)
+
+        for (String[] line : board) {                              // 2차원 배열을 "."로 채운다.
+            Arrays.fill(line, ".");
+        }
```
지금 단계에선 불필요한 구현으로 생각되는데요, 어떤가요?

wheejuni commented on src/main/java/chess/pieces/Pawn.java:
```diff
@@ -1,20 +1,32 @@
 
     Pawn() {
-        this(Pawn.WHITE); // 생성자에서 같은 클래스의 다른 생성자를 호출한다.
+        this(Pawn.WHITE_COLOR, Pawn.WHITE_REPRESENTATION); // 생성자에서 같은 클래스의 다른 생성자를 호출한다.
```
아무런 이름이 없는 기본 생성자가 기본으로 흰색 폰을 만드는건 조금 어색합니다.
[static factory method](https://johngrib.github.io/wiki/static-factory-method-pattern/)에 대해서 학습해보시고, 적용해보시는 건 어떨까요?

ehdrhelr replied:
미션2에서 색이 없는 폰을 생성하는 경우 기본으로 흰색 폰을 생성한다는 요구사항을 그대로 계승해서 작성했습니다.
브라이언의 말을 듣고보니 요구사항이라도 어색한 느낌이 들고 제가 더 나은 방향에 대한 고민 없이 미션 구현에만 치중한 것이 아닐까 생각해보았습니다!
알려주신 static factory method를 적용해서 더 개선해보겠습니다~!

wheejuni commented on src/main/java/chess/pieces/Pawn.java:
```diff
@@ -1,20 +1,32 @@
 public class Pawn {
-    public static final String BLACK = "black";
-    public static final String WHITE = "white";
+    public static final String BLACK_COLOR = "black";
+    public static final String WHITE_COLOR = "white";
+    public static final String BLACK_REPRESENTATION = "P";
+    public static final String WHITE_REPRESENTATION = "p";
```
`enum` 으로 관리하는게 더 좋아보이는데요? 어떠세요?

ehdrhelr replied:
enum으로 관리하게 되면 데이터타입으로 상수들을 격리시켜 놓으니 Pawn 생성자의 인자로 사용할 때나, 어떤 검증에 사용될 때 안정성이 생기는 장점이 생기는 것 같습니다.
enum을 사용해서 더 개선해보겠습니다ㅎㅎㅎ

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -3,26 +3,114 @@
+    }
+
+    public String getWhitePawnsResult() {
+        StringBuilder whitePawnRepresentations = new StringBuilder("");
```
불필요한 생성자 호출입니다.

```suggestion
        StringBuilder whitePawnRepresentations = new StringBuilder();

ehdrhelr replied:
StringBuilder 생성자에 대해서 더 공부해보겠습니다!

wheejuni reviewed (changes requested):
3단계까지 진행하시느라 정말 수고 많으셨습니다. 💯 
다만 지금 단계에서 작성하신 코드에 대해 궁금한 점도 있고, 아직 merge가 되기에는 부족한 점들이 있는 듯합니다.
많은 고민을 하시다보니 코드가 다소 복잡해지고 오버스펙이 된 면이 있는 듯합니다. 

코드에 대한 의문점과 개선 제안사항을 코멘트로 남겨드렸으니 확인해 주시고, 대댓글 혹은 수정 부탁드려요.

Commit:
refactor : not use two-dimensional array to represent board

1. initialize() only saves pawns in their lists.

2. print() prints out Board's initial status which Board manages by itself.

3. Remove literals and methods got useless as per above changes.

4. Test codes reflect above changes.

5. Remove an useless parameter("") in StringBuilder's constructor.

Commit:
refactor : use static factory method in Pawn class

1. Change Pawn's constructors to it's static factory methods where they are used.

2. Test codes changed by using them.

Commit:
refactor : Add enum Color

1. Add removeAll in initialize() to make it more meaningful.

ghost commented:
리뷰를 참고하여 개선 후 커밋하였으니 참고해주시면 감사하겠습니다~!

wheejuni reviewed (approved):
👍
